### PR TITLE
Rename prim to host

### PR DESCRIPTION
**Pull request by AstroSnail**
ontop of #95 which means merging this will merge that too. take care if you want to review #95 before merging it

use this link to see the changes particular to this pr https://github.com/AstroSnail/Alicorn0/compare/feature/refactor-terms-gen...feature/rename-prim-to-host

### CODE REVIEW

File: alicorn-expressions.lua [hunk 1]
```diff
@@ -1109,35 +1110,27 @@ local block = metalanguage.reducer(
 	"block"
 )
 
--- example usage of primitive_applicative
```
Comment (ErikMcClure) [inline alicorn-expressions.lua:1112]: this example was deleted, is it no longer applicable?

File: derivers.lua [hunk 2]
```diff
@@ -143,8 +237,187 @@ local is = {
 			assert(compiled, message)
 			idx["is_" .. vname] = compiled()
 		end
+
+		t.derived_is = true
+	end,
+	foreign = function()
+		error("can't derive is for a foreign type")
```
Comment (ErikMcClure) [inline derivers.lua:244]: should edit this to say "can't derive is() for a foreign type" to make it more clear

File: derivers.lua [hunk 3]
```diff
@@ -143,8 +237,187 @@ local is = {
...
+
+			local compiled, message = load(chunk, "derive-unwrap_enum", "t")
+			assert(compiled, message)
+			idx["unwrap_" .. vname] = compiled()
+		end
+
+		t.derived_unwrap = true
+	end,
+	foreign = function()
+		error("can't derive unwrap for a foreign type")
```
Comment (ErikMcClure) [inline derivers.lua:340]: change to "can't derive unwrap()" for more clarity

File: derivers.lua [hunk 4]
```diff
@@ -143,8 +237,187 @@ local is = {
...
+
+			local compiled, message = load(chunk, "derive-as_enum", "t")
+			assert(compiled, message)
+			idx["as_" .. vname] = compiled()
+		end
+
+		t.derived_as = true
+	end,
+	foreign = function()
+		error("can't derive as for a foreign type")
```
Comment (ErikMcClure) [inline derivers.lua:410]: change to "can't derive as(0" for more clarity

File: derivers.lua [hunk 5]
```diff
@@ -511,7 +809,145 @@ local diff = {
 			variants_checks[left.kind](left, right)
 		end
 
-		idx["diff"] = diff_fn
+		idx.diff = diff_fn
+		t.derived_diff = true
+	end,
+	foreign = function()
+		error("can't derive diff for a foreign type")
```
Comment (ErikMcClure) [inline derivers.lua:816]: change to "can't derive diff()" for more clarity

File: test-syntax-to-terms-to-eval.lua
Comment (ErikMcClure) [inline test-syntax-to-terms-to-eval.lua:1]: possibly out of scope of this PR: We need to move the tests we are using to a tests folder with a way to run them all.

**Review by ErikMcClure — APPROVED** [review]
Some nitpicks but this can be merged and nitpicks fixed later if it is blocking. I would like to actually organize our tests as well, but that could be in a separate PR.

File: alicorn-expressions.lua (hunk 1 above)
Comment (AstroSnail) [reply alicorn-expressions.lua]: thought they were unused, turns out some of them are still hanging on and apparently not broken

File: test-syntax-to-terms-to-eval.lua
Comment (AstroSnail) [reply test-syntax-to-terms-to-eval.lua:1]: i will need guidance on how to run some of the tests, and some will need to be improved with clear indications of pass/fail. probably should go in another pr

File: derivers.lua (hunk 2 above)
Comment (AstroSnail) [reply derivers.lua:244]: resolved in #95

File: derivers.lua (hunk 3 above)
Comment (AstroSnail) [reply derivers.lua:340]: resolved in #95

File: derivers.lua (hunk 4 above)
Comment (AstroSnail) [reply derivers.lua:410]: resolved in #95

File: derivers.lua (hunk 5 above)
Comment (AstroSnail) [reply derivers.lua:816]: resolved in #95

**Review by aiverson — APPROVED** [review]

**Pull request merged**